### DISCUSSION TITLE
ipopt & c-code generation clarification

### DISCUSSION THREAD

**Discussion by goosst** [casadi/casadi#3947] [Q&A] [answered]
Hello

I'm a bit confused if c-code generated with ipopt as solver is supported. Based on the documentation and other posts in this discussion forum, I'm not sure. 
So before losing many hours chasing my tail, asking the question what the recommended way of using the generated c-code with ipopt is? 
Does it need to be called with the c-api and ipopt related items will just work https://web.casadi.org/blog/mpc-simulink2/?

In my case, C-code was generated through the "example text book" opti.to_function and then generating c-code.

```
opti.solver('ipopt');
 
....

% Create CasADi function for C code generation
c_function = opti.to_function('optimal_control', {state, outdoortemp, setpoint}, {watersetpoints(1), valvesetpoints(1)});

% Set options for C code generation, including main and verbose
opts = struct();
opts.main = true;       % Include a main function
opts.verbose = true;    % Display verbose output during code generation
opts.with_header=false;

% Generate C code with the specified options
c_function.generate('optimal_control.c', opts);
```

Compiling as a shared library goes without problems (did not try to use it yet), compiling it as an executable seems to fail. 

%str_compile=strcat('gcc -fPIC -shared -O2 -I',str_include,' -L',DIR,'-lcasadi -lipopt optimal_control.c -o liboptimal_control.so')
str_compile=strcat('gcc -O2 -I',str_include,' -L',DIR,'-lcasadi -lipopt -lm optimal_control.c -o optimal_control')

```
/usr/bin/ld: /tmp/ccwEOmHc.o: in function `optimal_control_fmax':
optimal_control.c:(.text+0x17d7): undefined reference to `fmax'
/usr/bin/ld: /tmp/ccwEOmHc.o: in function `optimal_control_ipopt_presolve':
optimal_control.c:(.text+0x1fa3): undefined reference to `CreateIpoptProblem'
/usr/bin/ld: /tmp/ccwEOmHc.o: in function `optimal_control_ipopt_solve':
optimal_control.c:(.text+0x211d): undefined reference to `IpoptSolve'
/usr/bin/ld: optimal_control.c:(.text+0x219d): undefined reference to `FreeIpoptProblem'
/usr/bin/ld: /tmp/ccwEOmHc.o: in function `optimal_control_f1':
optimal_control.c:(.text+0x139c7d): undefined reference to `AddIpoptStrOption'
collect2: error: ld returned 1 exit status
```

**Comment by jgillis** [comment, marked as answer]
Generated C code with an Ipopt solver embedded is supported since 3.6.5; you'll probably find many older statements that it was impossible.
You should link with Ipopt (`-lipopt` and `-L<DIR>` with DIR coming from `GlobalOptions.getCasadiPath()` ).
The `fmax` thing should be resolved with `-lm`.

Your command is very close to functional in fact. Just put the c file first and the linker flags later.

**Comment by goosst** [comment]
thank you, with some extra arguments I've got an application which executes the optimization :). 

DIR=GlobalOptions.getCasadiPath();
str_include = GlobalOptions.getCasadiIncludePath();
str_compile=strcat('gcc -L',DIR,' -Wl,-rpath,',DIR,' -I',str_include,' optimal_control.c -lm -lipopt -o optimal_control')
